controllers/machinepool_controller.go

```diff
@@ -57,13 +76,181 @@ func (r *MachinePoolReconciler) SetupWithManager(mgr ctrl.Manager, options contr
 	r.controller = c
 	r.recorder = mgr.GetEventRecorderFor("machinepool-controller")
 	r.config = mgr.GetConfig()
-
 	r.scheme = mgr.GetScheme()
 	return nil
 }
 
-func (r *MachinePoolReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	// TODO(juan-lee): Add machine pool implementation.
+func (r *MachinePoolReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr error) {
 	// TODO(vincepri): Add support for pause annotation and honoring Cluster.Spec.Paused.
+	ctx := context.Background()
+	logger := r.Log.WithValues("machinepool", req.NamespacedName)
+
+	mp := &clusterv1.MachinePool{}
+	if err := r.Client.Get(ctx, req.NamespacedName, mp); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return. Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	cluster, err := util.GetClusterByName(ctx, r.Client, mp.ObjectMeta.Namespace, mp.Spec.ClusterName)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to get cluster %q for machine %q in namespace %q",
+			mp.Spec.ClusterName, mp.Name, mp.Namespace)
+	}
+
+	// Return early if the object or Cluster is paused.
+	if util.IsPaused(cluster, mp) {
+		logger.V(3).Info("reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(mp, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		r.reconcilePhase(mp)
+		// TODO(jpang): add support for metrics.
+
+		// Always attempt to patch the object and status after each reconciliation.
+		if err := patchHelper.Patch(ctx, mp); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
+	// Reconcile labels.
+	if mp.Labels == nil {
+		mp.Labels = make(map[string]string)
+	}
+	mp.Labels[clusterv1.ClusterLabelName] = mp.Spec.ClusterName
+
+	// Handle deletion reconciliation loop.
+	if !mp.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, cluster, mp)
+	}
+
+	// Handle normal reconciliation loop.
+	return r.reconcile(ctx, cluster, mp)
+}
+
+func (r *MachinePoolReconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster, mp *clusterv1.MachinePool) (ctrl.Result, error) {
+	// If the MachinePool belongs to a cluster, add an owner reference.
```

Comment:
As discussed in #2101, we can probably omit the call to shouldAdopt and always ensure there is an owner ref to the cluster.